### PR TITLE
Fix crash due to stale pointers in font cache after ass_set_fonts

### PR DESCRIPTION
**Pull request by rcombs**
I'm kinda confused as to why this doesn't cause more problems.

Also some warning fixes.

### CODE REVIEW

File: libass/ass_fontselect.c [hunk 1]
```diff
@@ -50,6 +50,11 @@
 #define ABS(x) ((x) < 0 ? -(x) : (x))
 #define MAX_FULLNAME 100
 
+// workaround for gcc and clang bugs
+// https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80454
+// https://bugs.llvm.org/show_bug.cgi?id=21689
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
```
Comment (wm4) [inline libass/ass_fontselect.c]: I'd rather just put it as normal flag into the configure script.

File: libass/ass_render_api.c [hunk 2]
```diff
@@ -150,6 +150,8 @@ void ass_set_fonts(ASS_Renderer *priv, const char *default_font,
         ass_fontselect_free(priv->fontselect);
     priv->fontselect = ass_fontselect_init(priv->library, priv->ftlibrary,
             default_family, default_font, config, dfp);
+
+    ass_cache_empty(priv->cache.font_cache);
```
Comment (wm4) [inline libass/ass_render_api.c]: Maybe better call it before the dependent objects are destroyed?

**Comment by MrSmile** [comment]
Actually, `ass_cache_empty()` does not delete all cached `ASS_Font` objects. Such objects can be kept alive by references in keys of other caches. So it's necessary to either drop all caches altogether or somehow circumvent orphan fonts.

**Comment by rcombs** [comment]
Hmmm, the outline cache and the glyph metrics cache, I guess?

**Comment by MrSmile** [comment]
Also bitmap and composite due to chain dependency (i.e. literally all caches). But all except glyph metrics are already handled in `ass_reconfigure()`, so only one left.

**Comment by wm4** [comment]
This is complicated by the fact that we probably want to eventually allow the user to keep bitmaps for a longer time via refcounting.

**Comment by MrSmile** [comment]
Indeed. It's necessary to split cache key and value destruction in the future. That way alive bitmaps won't keep all related cache hierarchy from clearing.

**Comment by rcombs** [comment]
Am I still missing anything else that needs clearing here? And are we good with the other 2 changes?

**Comment by wm4** [comment]
LGTM I guess.

**Comment by astiob** [comment]
I guess it may be worth noting in the commit message that it fixes a crash, but generally LGTM.

**Comment by astiob** [comment]
I see the cache-related commit is now in master as 68f25e2c26de2c5c3624bf0fe6d12a0e9c35e861, so shall we close this now or keep it open for the warning-related commit?

In the warning-related commit, it may be good to note in the commit message that the warning is being disabled specifically because Clang gives it for the universal initializer `={0}`. I presume we could revert this change whenever Clang stops doing that.

**Comment by astiob** [comment]
I’ve pushed the `={0}` warning fix. I believe the `ASS_Outline` warning was fixed by @MrSmile’s last merged PR, so there’s nothing else left here.

**Pull request closed without merging**